src/app/modules/shared_models/message_reaction_model.nim

```diff
@@ -0,0 +1,146 @@
+    {
+      ModelRole.EmojiId.int:"emojiId",
+      ModelRole.AmIReactedWithThisEmoji.int:"amIReactedWithThisEmoji",
+      ModelRole.NumberOfReactions.int:"numberOfReactions",
+      ModelRole.JsonArrayOfUsersReactedWithThisEmoji.int: "jsonArrayOfUsersReactedWithThisEmoji"
+    }.toTable
+
+  method data(self: MessageReactionModel, index: QModelIndex, role: int): QVariant =
+    if (not index.isValid):
+      return
+
+    if (index.row < 0 or index.row >= self.items.len):
+      return
+
+    let item = self.items[index.row]
+    let enumRole = role.ModelRole
+
+    case enumRole:
+    of ModelRole.EmojiId: 
+      result = newQVariant(item.emojiId.int)
+    of ModelRole.AmIReactedWithThisEmoji: 
+      result = newQVariant(item.amIReactedWithThisEmoji)
+    of ModelRole.NumberOfReactions: 
+      result = newQVariant(item.numberOfReactions)
+    of ModelRole.JsonArrayOfUsersReactedWithThisEmoji: 
+      # Would be good if we could return QVariant of array (seq) here, but it's not supported in our NimQml, 
+      # because of that we're returning json array as a string.
+      result = newQVariant($item.jsonArrayOfUsersReactedWithThisEmoji)
+
+  proc reactionItemWithEmojiIdExists(self: MessageReactionModel, emojiId: EmojiId): bool = 
+    for it in self.items:
+      if(it.emojiId == emojiId):
+        return true
+    return false
+
+  proc getIndexOfTheItemWithEmojiId(self: MessageReactionModel, emojiId: EmojiId): int = 
+    for i in 0..<self.items.len:
+      if(self.items[i].emojiId == emojiId):
+        return i
+    return -1
+
+  proc findPositionForTheItemWithEmojiId(self: MessageReactionModel, emojiId: EmojiId): int = 
+    if(self.items.len == 0):
+      return 0
+
+    for i in 0..<self.items.len:
+      if(emojiId < self.items[i].emojiId):
+        return i
+
+    return self.items.len
+
+  proc shouldAddReaction*(self: MessageReactionModel, emojiId: EmojiId, userPublicKey: string): bool = 
+    let ind = self.getIndexOfTheItemWithEmojiId(emojiId)
+    if(ind == -1):
+      return true
+    return self.items[ind].shouldAddReaction(userPublicKey)
+
+  proc getReactionId*(self: MessageReactionModel, emojiId: EmojiId, userPublicKey: string): string = 
+    let ind = self.getIndexOfTheItemWithEmojiId(emojiId)
+    if(ind == -1):
+      return ""
+    return self.items[ind].getReactionId(userPublicKey)
+
+  proc addReaction*(self: MessageReactionModel, emojiId: EmojiId, amIReactedWithThisEmoji: bool, userPublicKey: string, 
+    userDisplayName: string, reactionId: string) = 
+    if(self.reactionItemWithEmojiIdExists(emojiId)):
+      let ind = self.getIndexOfTheItemWithEmojiId(emojiId)
+      if(ind == -1):
+        return
+      self.items[ind].addReaction(amIReactedWithThisEmoji, userPublicKey, userDisplayName, reactionId)
+      let index = self.createIndex(ind, 0, nil)
+      self.dataChanged(index, index, @[]) # all roles
+    else:
+      let parentModelIndex = newQModelIndex()
+      defer: parentModelIndex.delete
```

Comment:
What does this do?

Reply:
@jrainville if you refer to this line `defer: parentModelIndex.delete`, it actually does deferred deletion of `newQModelIndex` and avoids memory leak caused by creating `new QModelIndex` on the c++ side.